### PR TITLE
Stabilize #[doc_alias = "..."]

### LINKED ISSUE #63222: [rustdoc] stabilize #[doc(alias = ...)]
cc @rust-lang/rustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #63222

cc @rust-lang/rustdoc 
r? @ollie27

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Hum, interesting issue: we need `doc_alias` to be stabilized before being able to stabilize it. Is there any way to not fail tests in such case?

cc @rust-lang/compiler

**Comment by Mark-Simulacrum** [comment]
`#[cfg_attr(bootstrap, doc(alias(...))]` is the way to go to avoid the boostrap compiler complaining about the feature not being stabilized.

I would also expect stabilization PRs to come with a report of the usage of the feature, even if brief -- how long has it been unstable for, what is the usage like in the wild. In particular, in this case it seems like this feature is missing documentation about the namespace we're aliasing items in the documentation to. If this is purely for search, then naming it simply "alias" feels too broad.

**Comment by Centril** [comment]
> I would also expect stabilization PRs to come with a report of the usage of the feature, even if brief -- how long has it been unstable for, what is the usage like in the wild. In particular, in this case it seems like this feature is missing documentation about the namespace we're aliasing items in the documentation to. If this is purely for search, then naming it simply "alias" feels too broad.

I think it's up to the `rustdoc` team to decide how they want to do stabilizations. They do not have to follow how the language team does things. But in case they want to do that, here's a sample stabilization report: https://github.com/rust-lang/rust/pull/61682#issuecomment-502472847.  I agree it's good in general to have such reports (maybe not as long tho... ^^).

**Comment by crlf0710** [comment]
Admittedly i've not have any experience on this, but how does this function works? I mean, if in a crate X there a Foo that aliased to Y, does this mean when i search either `X::Y` or `Y`, `X::Foo` will be given high priority in the search result?

**Comment by GuillaumeGomez** [comment]
@crlf0710 Aliases work at the crate level. So if you can a crate `A` providing a type `A::Foo` and a crate `B` aliasing a type to `Foo` as well, then the search will return both results. You can already check how aliases work by looking for "\[\]" in the std docs for example.

**Comment by JohnTitor** [comment]
Ping from triage: @ollie27 @GuillaumeGomez any updates on this?

**Comment by GuillaumeGomez** [comment]
I still need to check how to be able to actually stabilize this since we need it to be stable to stabilize it (I love paradoxes).

**Comment by joelpalmer** [comment]
Ping from Triage: Hi @GuillaumeGomez - what is the status of your paradox?

**Comment by GuillaumeGomez** [comment]
Didn't check yet (going through my other PRs for the moment).

**Comment by JohnCSimon** [comment]
Ping from triage 
@GuillaumeGomez Can you please resolve the merge conflict? 

Thank you!

**Comment by GuillaumeGomez** [comment]
The merge conflict is minor. What I need to do is fix the paradox issue first. Will do as soon as I have enough time!

**Comment by JohnCSimon** [comment]
Ping from triage:
@GuillaumeGomez Can you please address the merge conflicts? 
Thank you.

**Comment by GuillaumeGomez** [comment]
Waiting for approval first. A bit annoying to keep updating whereas there is no review...

**Comment by QuietMisdreavus** [comment]
Does anyone want to use this outside the standard library? I was under the impression that the purpose of this feature was to search for the operator traits via the operator itself. I could see this being used to associate a type with a common misnomer for it, but i'm not sure if anyone has thought to properly encode something like that. I'm just not sure this feature will see much use outside the standard library, and if that's the case we should just keep it perma-unstable.

As for your "paradox", @Mark-Simulacrum [already solved it for you](https://github.com/rust-lang/rust/pull/63236#issuecomment-517920130). Replace the `#![feature(doc_alias)]` attributes you deleted with `#![cfg_attr(bootstrap, feature(doc_alias))]` instead. That will keep it declared for the bootstrap compiler, which is the one that's complaining, but take it out in the final one, since that one will have declared it stable.

**Comment by GuillaumeGomez** [comment]
@QuietMisdreavus Yes, @Mark-Simulacrum told me but thanks for the update! As for its usage outside of libstd, I'd like to use it on gtk-rs (because we have to prepend some types with the lib name to avoid conflicts).

**Comment by QuietMisdreavus** [comment]
@GuillaumeGomez So you need to name the item something like `gtk_something` but you want to alias it to just `something` so people can find it easier? Wouldn't that already be taken care of by restricting the crate you search for and/or having our search logic favor items that contain the search string as an exact substring (which i thought we already did)? Am i understanding this wrong?

**Comment by GuillaumeGomez** [comment]
The naming won't change, even if you limit the search to a specific crate. For example, we have `GtkWindowExt` and I'd like to alias it to `WindowsExt`. We have a few examples like this. It's a detail at this point but something I'd like to have to.

However, for crates heavily using proc macros, it'd make much more sense since they can provide a completely new syntax.

**Comment by JohnTitor** [comment]
Ping from triage: @ollie27 could you review this?

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -854,6 +851,8 @@ declare_features! (
     (accepted, repr_align_enum, "1.37.0", Some(57996), None),
     // Allows `const _: TYPE = VALUE`.
     (accepted, underscore_const_names, "1.37.0", Some(54912), None),
+    // Allows `#[doc(alias = "...")]`.
+    (accepted, doc_alias, "1.27.0", Some(50146), None),
```
Comment (Centril) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    (accepted, doc_alias, "1.41.0", Some(50146), None),
```

Comment (GuillaumeGomez) [reply src/libsyntax/feature_gate.rs]: Sill waiting for approval before rebasing. It's just one of the conflicts. ;)

Comment (Centril) [reply src/libsyntax/feature_gate.rs]: Hehe; I just did this for all 1.40.0 PRs that were relevant ^^

Comment (GuillaumeGomez) [reply src/libsyntax/feature_gate.rs]: :p

**Comment by JohnTitor** [comment]
Ping from triage: assigning another reviewer randomly...
r? @QuietMisdreavus
CC: @rust-lang/rustdoc

**Comment by ollie27** [comment]
I think this should be closed for now to keep the PR queue clean. It needs a FCP anyway which can happen on the tracking issue (#50146).

**Comment by GuillaumeGomez** [comment]
@ollie27 Ok, then please start the FCP on the issue. :)

**Comment by ollie27** [comment]
> 
> 
> Ok, then please start the FCP on the issue. :)

@GuillaumeGomez as you're the one proposing stabilization it would be better if you propose the FCP. Make sure you include an explanation as to why you think it should be stabilized and more specifically what it would be used for outside of `std`. That's what was missing from this PR.

**Comment by GuillaumeGomez** [comment]
Sure!

**Pull request closed without merging**